### PR TITLE
Sync review old protocol

### PR DESCRIPTION
**Pull request by lsebrie**
- only timed status messages
- block relay only sends block hash
- mined block is the only one sent by default
- missing hashes are searched only when a mined block arrives or a a get block message was sent earlier

### CODE REVIEW

**Comment by rskops** [comment]
SonarQube analysis reported 190 issues
* ![BLOCKER][BLOCKER] 1 blocker
* ![CRITICAL][CRITICAL] 11 critical
* ![MAJOR][MAJOR] 44 major
* ![MINOR][MINOR] 106 minor
* ![INFO][INFO] 28 info

#### Top 10 issues

1. ![BLOCKER][BLOCKER] [MinerServerImpl.java#L524](https://github.com/rsksmart/rskj/blob/ce287d67e37c00c7db58cdd5310f5b738fc017e1/rskj-core/src/main/java/co/rsk/mine/MinerServerImpl.java#L524): Catch Exception instead of Throwable. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1181)
1. ![CRITICAL][CRITICAL] [MinerClientImpl.java#L120](https://github.com/rsksmart/rskj/blob/ce287d67e37c00c7db58cdd5310f5b738fc017e1/rskj-core/src/main/java/co/rsk/mine/MinerClientImpl.java#L120): Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2142)
1. ![CRITICAL][CRITICAL] [MinerClientImpl.java#L128](https://github.com/rsksmart/rskj/blob/ce287d67e37c00c7db58cdd5310f5b738fc017e1/rskj-core/src/main/java/co/rsk/mine/MinerClientImpl.java#L128): Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2142)
1. ![CRITICAL][CRITICAL] [MinerClientImpl.java#L140](https://github.com/rsksmart/rskj/blob/ce287d67e37c00c7db58cdd5310f5b738fc017e1/rskj-core/src/main/java/co/rsk/mine/MinerClientImpl.java#L140): Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2142)
1. ![CRITICAL][CRITICAL] [TxBuilderEx.java#L72](https://github.com/rsksmart/rskj/blob/ce287d67e37c00c7db58cdd5310f5b738fc017e1/rskj-core/src/main/java/co/rsk/mine/TxBuilderEx.java#L72): Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2142)
1. ![CRITICAL][CRITICAL] [TxBuilderEx.java#L79](https://github.com/rsksmart/rskj/blob/ce287d67e37c00c7db58cdd5310f5b738fc017e1/rskj-core/src/main/java/co/rsk/mine/TxBuilderEx.java#L79): Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2142)
1. ![CRITICAL][CRITICAL] [TxBuilderEx.java#L87](https://github.com/rsksmart/rskj/blob/ce287d67e37c00c7db58cdd5310f5b738fc017e1/rskj-core/src/main/java/co/rsk/mine/TxBuilderEx.java#L87): Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2142)
1. ![CRITICAL][CRITICAL] [TxBuilderEx.java#L99](https://github.com/rsksmart/rskj/blob/ce287d67e37c00c7db58cdd5310f5b738fc017e1/rskj-core/src/main/java/co/rsk/mine/TxBuilderEx.java#L99): Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2142)
1. ![CRITICAL][CRITICAL] [TxBuilderEx.java#L127](https://github.com/rsksmart/rskj/blob/ce287d67e37c00c7db58cdd5310f5b738fc017e1/rskj-core/src/main/java/co/rsk/mine/TxBuilderEx.java#L127): Cast one of the operands of this addition operation to a "long". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2184)
1. ![CRITICAL][CRITICAL] [TxBuilderEx.java#L131](https://github.com/rsksmart/rskj/blob/ce287d67e37c00c7db58cdd5310f5b738fc017e1/rskj-core/src/main/java/co/rsk/mine/TxBuilderEx.java#L131): Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2142)

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: rskj-core/src/main/java/co/rsk/net/NodeBlockProcessor.java [hunk 1]
```diff
@@ -45,28 +44,24 @@
 public class NodeBlockProcessor implements BlockProcessor {
     private static final Logger logger = LoggerFactory.getLogger("blockprocessor");
 
-    private final Object statusLock = new Object();
-
     private final BlockStore store;
     private final Blockchain blockchain;
-    private final BlockNodeInformation nodeInformation; // keep tabs on which nodes know which blocks.
+    // keep tabs on which nodes know which blocks.
```
Comment (tinchou) [inline rskj-core/src/main/java/co/rsk/net/NodeBlockProcessor.java]: The comment is now in the wrong place.

File: rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java [hunk 2]
```diff
@@ -247,20 +239,12 @@ private synchronized void sendStatusToAll() {
         this.channelManager.broadcastStatus(status);
     }
 
-    @CheckForNull
     public synchronized Block getBestBlock() {
-        if (this.blockProcessor != null)
-            return this.blockProcessor.getBlockchain().getBestBlock();
-
-        return null;
+        return this.blockProcessor.getBlockchain().getBestBlock();
```
Comment (tinchou) [inline rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java]: Eventually, we'll want a direct dependency on Blockchain :)

File: rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java [hunk 3]
```diff
@@ -297,202 +281,122 @@ private void processBlockMessage(@Nonnull final MessageChannel sender, @Nonnull
...
-            this.blockProcessor.acceptAnyBlock();
-            this.receivedMessages.clear();
-        }
+        Metrics.processBlockMessage("finish", block, sender.getPeerNodeID());
+    }
 
+    private void tryRelayBlock(@Nonnull MessageChannel sender, Block block, boolean wasOrphan, BlockProcessResult result) {
         // is new block and it is not orphan, it is in some blockchain
-        if (wasOrphan && result.wasBlockAdded(block) && !this.blockProcessor.isSyncingBlocks())
+        if (!wasOrphan && result.wasBlockAdded(block) && !this.blockProcessor.hasBetterBlockToSync())
```
Comment (tinchou) [inline rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java]: Looks like we had a bug, great catch

File: rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java [hunk 4]
```diff
@@ -505,42 +409,31 @@ private void processTransactionsMessage(@Nonnull final MessageChannel sender, @N
...
+            final ByteArrayWrapper txHash = new ByteArrayWrapper(tx.getHash());
+            transactionNodeInformation.addTransactionToNode(txHash, sender.getPeerNodeID());
+            final Set<NodeID> nodesToSkip = new HashSet<>(transactionNodeInformation.getNodesByTransaction(tx.getHash()));
+            final Set<NodeID> newNodes = channelManager.broadcastTransaction(tx, nodesToSkip);
 
-        Metrics.processTxsMessage("txToNodeInfoUpdated", acceptedTxs, sender.getPeerNodeID());
-        Metrics.processTxsMessage("finish", acceptedTxs, sender.getPeerNodeID());
-
-        loggerMessageProcess.debug("Tx message process finished after [{}] nano.", System.nanoTime() - start);
+            for (NodeID nodeID : newNodes)
```
Comment (tinchou) [inline rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java]: braces

File: rskj-core/src/main/java/co/rsk/net/sync/DownloadingBodiesSyncState.java [hunk 5]
```diff
@@ -177,8 +177,11 @@ private void clearPeerInfo(NodeID peerId) {
 
     private Optional<BlockHeader> updateHeadersAndChunks(NodeID peerId, Integer currentChunk) {
         Deque<BlockHeader> headers = pendingHeaders.get(currentChunk);
-        if (!headers.isEmpty()) {
-            return Optional.of(headers.pop());
+        while (!headers.isEmpty()) {
+            BlockHeader header = headers.pop();
+            // we double check if the header was not downloaded or obtained by another way
+            if (!syncInformation.isKnownBlock(header.getHash()))
```
Comment (tinchou) [inline rskj-core/src/main/java/co/rsk/net/sync/DownloadingBodiesSyncState.java]: braces

File: rskj-core/src/main/java/co/rsk/net/sync/DownloadingBodiesSyncState.java [hunk 6]
```diff
@@ -193,20 +196,20 @@ private void clearPeerInfo(NodeID peerId) {
...
                 Deque<BlockHeader> headers = pendingHeaders.get(chunkNumber);
-                if (!headers.isEmpty()) {
-                    chunksBeingDownloaded.put(peerId, chunkNumber);
-                    segmentsBeingDownloaded.put(peerId, segmentNumber);
-                    return Optional.of(headers.poll());
-                } else {
-                    // log something we are in trouble
-                    syncEventsHandler.warnMessage("We got a segment {} with a unfinished chunk {} with no pending headers!", segmentNumber, chunkNumber);
-                    syncEventsHandler.stopSyncing();
+                while (!headers.isEmpty()) {
```
Comment (tinchou) [inline rskj-core/src/main/java/co/rsk/net/sync/DownloadingBodiesSyncState.java]: FYI this is not thread safe

File: rskj-core/src/main/java/org/ethereum/core/Transaction.java [hunk 7]
```diff
@@ -308,10 +308,14 @@ public ECDSASignature getSignature() {
     }
 
     public boolean acceptTransactionSignature() {
```
Comment (tinchou) [inline rskj-core/src/main/java/org/ethereum/core/Transaction.java]: braces

File: rskj-core/src/main/java/co/rsk/net/BlockProcessResult.java [hunk 8]
```diff
@@ -37,9 +37,11 @@
 
     private Map<ByteArrayWrapper, ImportResult> result;
 
-    public BlockProcessResult(boolean additionalValidations, Map<ByteArrayWrapper, ImportResult> result) {
+    public BlockProcessResult(boolean additionalValidations, Map<ByteArrayWrapper, ImportResult> result, String blockHash, long time) {
         this.additionalValidationsOk = additionalValidations;
         this.result = result;
+        if (time >= 1000000000)
```
[the remark this answers is not in the document]
Comment (lsebrie) [reply rskj-core/src/main/java/co/rsk/net/BlockProcessResult.java]: :+1:

File: rskj-core/src/main/java/co/rsk/net/BlockProcessResult.java [hunk 9]
```diff
@@ -37,9 +37,11 @@
 
     private Map<ByteArrayWrapper, ImportResult> result;
 
-    public BlockProcessResult(boolean additionalValidations, Map<ByteArrayWrapper, ImportResult> result) {
+    public BlockProcessResult(boolean additionalValidations, Map<ByteArrayWrapper, ImportResult> result, String blockHash, long time) {
```
[the remark this answers is not in the document]
Comment (lsebrie) [reply rskj-core/src/main/java/co/rsk/net/BlockProcessResult.java]: Improved :+1:

File: rskj-core/src/main/java/co/rsk/net/BlockStore.java [hunk 10]
```diff
@@ -149,15 +149,12 @@ public synchronized Block getBlockByHash(byte[] hash) {
     /**
      * getChildrenOf returns all the children of a list of blocks that are in the BlockStore.
      *
-     * @param blocks the list of blocks to retrieve the children.
+     * @param blocks a set of blocks to retrieve the children.
      * @return A list with all the children of the given list of blocks.
      */
-    public List<Block> getChildrenOf(List<Block> blocks) {
-        Set<Block> children = blocks.stream()
+    public List<Block> getChildrenOf(Set<Block> blocks) {
```
[the remark this answers is not in the document]
Comment (lsebrie) [reply rskj-core/src/main/java/co/rsk/net/BlockStore.java]: Yes, the change was not intended. I'll fix this

File: rskj-core/src/main/java/co/rsk/net/BlockSyncService.java [hunk 11]
```diff
@@ -115,53 +84,46 @@ public BlockProcessResult processBlock(MessageChannel sender, @Nonnull Block blo
...
-        }
+    private void tryReleaseStore(long bestBlockNumber) {
+        if ((++processedBlocksCounter % 200) == 0) {
+            long minimal = store.minimalHeight();
+            long maximum = store.maximumHeight();
+            logger.trace("Blocks in block processor {} from height {} to height {}", this.store.size(), minimal, maximum);
 
-        if (!syncing) {
-            nsyncs++;
+            if (minimal < bestBlockNumber - 1000)
```
[the remark this answers is not in the document]
Comment (lsebrie) [reply rskj-core/src/main/java/co/rsk/net/BlockSyncService.java]: :+1:

File: rskj-core/src/main/java/co/rsk/net/BlockSyncService.java [hunk 12]
```diff
@@ -115,53 +84,46 @@ public BlockProcessResult processBlock(MessageChannel sender, @Nonnull Block blo
...
-
-        return nsyncs < 2;
     }
 
     public boolean hasBetterBlockToSync() {
         long last = this.getLastKnownBlockNumber();
         long current = this.getBestBlockNumber();
 
-        return last >= current + NBLOCKS_TO_SYNC;
+        return last >= current + syncConfiguration.getChunkSize() / 8;
```
[the remark this answers is not in the document]
Comment (lsebrie) [reply rskj-core/src/main/java/co/rsk/net/BlockSyncService.java]: :+1:

File: rskj-core/src/main/java/co/rsk/net/NodeBlockProcessor.java (hunk 1 above)
Comment (lsebrie) [reply rskj-core/src/main/java/co/rsk/net/NodeBlockProcessor.java]: Moved :+1:

File: rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java (hunk 2 above)
Comment (lsebrie) [reply rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java]: Yes, we still have some dependencies to improve. This is a step.

File: rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java (hunk 3 above)
Comment (lsebrie) [reply rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java]: I read the code and I believe the proper implementation is as said on the comment line

File: rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java (hunk 4 above)
Comment (lsebrie) [reply rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java]: Improved :+1:

File: rskj-core/src/main/java/co/rsk/net/sync/DownloadingBodiesSyncState.java (hunk 6 above)
Comment (lsebrie) [reply rskj-core/src/main/java/co/rsk/net/sync/DownloadingBodiesSyncState.java]: Improved so if we need to change to a not thread safe execution it still works

File: rskj-core/src/main/java/co/rsk/net/sync/DownloadingBodiesSyncState.java (hunk 5 above)
Comment (lsebrie) [reply rskj-core/src/main/java/co/rsk/net/sync/DownloadingBodiesSyncState.java]: :+1:

File: rskj-core/src/main/java/org/ethereum/core/Transaction.java (hunk 7 above)
Comment (lsebrie) [reply rskj-core/src/main/java/org/ethereum/core/Transaction.java]: :+1:

**Comment by rskops** [comment]
SonarQube analysis reported 336 issues
* ![CRITICAL][CRITICAL] 11 critical
* ![MAJOR][MAJOR] 85 major
* ![MINOR][MINOR] 157 minor
* ![INFO][INFO] 83 info

#### Top 10 issues

1. ![CRITICAL][CRITICAL] [MinerClientImpl.java#L120](https://github.com/rsksmart/rskj/blob/478e0f4b5c347a8fee7f0ff3b398632e99f5925d/rskj-core/src/main/java/co/rsk/mine/MinerClientImpl.java#L120): Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2142)
1. ![CRITICAL][CRITICAL] [MinerClientImpl.java#L128](https://github.com/rsksmart/rskj/blob/478e0f4b5c347a8fee7f0ff3b398632e99f5925d/rskj-core/src/main/java/co/rsk/mine/MinerClientImpl.java#L128): Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2142)
1. ![CRITICAL][CRITICAL] [MinerClientImpl.java#L140](https://github.com/rsksmart/rskj/blob/478e0f4b5c347a8fee7f0ff3b398632e99f5925d/rskj-core/src/main/java/co/rsk/mine/MinerClientImpl.java#L140): Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2142)
1. ![CRITICAL][CRITICAL] [TxBuilderEx.java#L72](https://github.com/rsksmart/rskj/blob/478e0f4b5c347a8fee7f0ff3b398632e99f5925d/rskj-core/src/main/java/co/rsk/mine/TxBuilderEx.java#L72): Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2142)
1. ![CRITICAL][CRITICAL] [TxBuilderEx.java#L79](https://github.com/rsksmart/rskj/blob/478e0f4b5c347a8fee7f0ff3b398632e99f5925d/rskj-core/src/main/java/co/rsk/mine/TxBuilderEx.java#L79): Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2142)
1. ![CRITICAL][CRITICAL] [TxBuilderEx.java#L87](https://github.com/rsksmart/rskj/blob/478e0f4b5c347a8fee7f0ff3b398632e99f5925d/rskj-core/src/main/java/co/rsk/mine/TxBuilderEx.java#L87): Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2142)
1. ![CRITICAL][CRITICAL] [TxBuilderEx.java#L99](https://github.com/rsksmart/rskj/blob/478e0f4b5c347a8fee7f0ff3b398632e99f5925d/rskj-core/src/main/java/co/rsk/mine/TxBuilderEx.java#L99): Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2142)
1. ![CRITICAL][CRITICAL] [TxBuilderEx.java#L127](https://github.com/rsksmart/rskj/blob/478e0f4b5c347a8fee7f0ff3b398632e99f5925d/rskj-core/src/main/java/co/rsk/mine/TxBuilderEx.java#L127): Cast one of the operands of this addition operation to a "long". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2184)
1. ![CRITICAL][CRITICAL] [TxBuilderEx.java#L131](https://github.com/rsksmart/rskj/blob/478e0f4b5c347a8fee7f0ff3b398632e99f5925d/rskj-core/src/main/java/co/rsk/mine/TxBuilderEx.java#L131): Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2142)
1. ![CRITICAL][CRITICAL] [TxBuilderEx.java#L155](https://github.com/rsksmart/rskj/blob/478e0f4b5c347a8fee7f0ff3b398632e99f5925d/rskj-core/src/main/java/co/rsk/mine/TxBuilderEx.java#L155): Cast one of the operands of this addition operation to a "long". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2184)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**